src/parser.cpp

```diff
@@ -106,6 +120,273 @@ Buffer Usage
+    //
+    // drop support trailers/chunk-ext, use internal definition
+    //
+    // last-chunk = 1*"0" CRLF CRLF
+
+    if( buffers::buffer_size(input.data()) <
+        last_chunk_len )
+    {
+        ec = error::need_data;
+        return;
+    }
+
+    auto cbs = input.data();
+    discontiguous_iterator pos = cbs.begin();
+    discontiguous_iterator end = cbs.end();
+
+    std::size_t len = 0;
+    for( ; pos != end && *pos == '0'; ++pos, ++len )
+    {
+    }
+
+    if( len == 0 )
+    {
+        ec = error::bad_payload;
+        return;
+    }
+
+    std::size_t const close_len = 4; // for \r\n\r\n
+    if( buffers::buffer_size(input.data()) - len <
+        close_len )
+    {
+        ec = error::need_data;
+        return;
+    }
+
+    char tmp[close_len] = {};
+    for( std::size_t i = 0; i < close_len; ++i )
+        tmp[i] = *pos++;
+
+    core::string_view s(tmp, close_len);
+    if( s != "\r\n\r\n" )
+    {
+        ec = error::bad_payload;
+        return;
+    }
+
+    input.consume(len + close_len);
+};
+
+template <class DynamicBuffer>
+bool
+parse_chunked(
+    buffers::circular_buffer& input,
+    DynamicBuffer& output,
+    system::error_code& ec,
+    std::size_t& chunk_remain_,
+    bool& needs_chunk_close_)
+{
+    if( input.size() == 0 )
+    {
+        ec = error::need_data;
+        return false;
+    }
+
+    for(;;)
+    {
+        if( chunk_remain_ == 0 )
+        {
+            if( needs_chunk_close_ )
+            {
+                if( input.size() < 2 )
+                {
+                    ec = error::need_data;
+                    return false;
+                }
+
+                std::size_t const crlf_len = 2;
+                char tmp[crlf_len] = {};
+
+                buffers::buffer_copy(
+                    buffers::mutable_buffer(
+                        tmp, crlf_len),
+                    input.data());
+
+                core::string_view str(tmp, crlf_len);
+                if( str != "\r\n" )
+                {
+                    ec = error::bad_payload;
+                    return false;
+                }
+
+                input.consume(crlf_len);
+                needs_chunk_close_ = false;
+                continue;
+            }
+
+            parse_chunk_header(input, ec, chunk_remain_);
+            if( ec )
+            {
+                system::error_code ec2;
+                parse_last_chunk(input, ec2);
+                if( ec2 )
+                {
+                    if( ec2 == condition::need_more_input )
+                        ec = ec2;
+                    return false;
+                }
+
+                // complete
+                ec.clear();
+                return true;
+            }
+
+            needs_chunk_close_ = true;
+        }
+
+        if( input.size() == 0 )
+        {
+            ec = error::need_data;
+            return false;
+        }
+
+        if( output.capacity() == 0 )
+            detail::throw_length_error();
```

Comment:
Yeah, this is actually a discussion point I wanted to go over with you. This can theoretically happen, and I'm not sure how to handle it. So for now, I just throw because solving the interface design here is non-trivial, but doesn't necessarily need to be blocking.

I think I'll just add a TODO here for now, and we'll revisit this once we have more experience with the library being deployed in http-io.